### PR TITLE
Add heartbeat metric for DAG processor 

### LINKED ISSUE #42380: Add heartbeat metric for DAG processor
### Description

The scheduler emits a `scheduler_heartbeat`, which makes it easy to determine if the scheduler is up and running. We should add an equivalent for the DAG processor.

### Use case/motivation

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by rawwar**
closes: #42380

### CODE REVIEW

**Comment by rawwar** [comment]
I've tested it locally with a standalone dag processor and its emitting the heartbeat for dag processor:
image

File: airflow/jobs/dag_processor_job_runner.py [hunk 1]
```diff
@@ -52,7 +56,7 @@ def __init__(
         self.processor = processor
         self.processor.heartbeat = lambda: perform_heartbeat(
             job=self.job,
-            heartbeat_callback=empty_callback,
```
Comment (jedcunningham) [inline airflow/jobs/dag_processor_job_runner.py:55]: We can probably get rid of empty_callback too?

Comment (rawwar) [reply airflow/jobs/dag_processor_job_runner.py:55]: Removed it. Thanks!

**Review by jedcunningham — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Comment by shalberd** [comment]
@potiuk will this make it into 2.10.5 tagged branch?
Dag processor has been able to separately run for a while now in Airflow 2.x, yet I cannot find this metrics export change in any release notes. Is there a particular reason for it missing in e.g. 2.10.5?

**Comment by rawwar** [comment]
> @potiuk @rawwar will this make it into 2.10.5 tagged branch? Dag processor has been able to separately run for a while now in Airflow 2.x, yet I cannot find this metrics export change in any release notes. Is there a particular reason for it missing in e.g. 2.10.5?

@ephraimbuddy , @jedcunningham , can this be backported?

**Comment by ephraimbuddy** [comment]
> > @potiuk @rawwar will this make it into 2.10.5 tagged branch? Dag processor has been able to separately run for a while now in Airflow 2.x, yet I cannot find this metrics export change in any release notes. Is there a particular reason for it missing in e.g. 2.10.5?
> 
> @ephraimbuddy , @jedcunningham , can this be backported?

No. This is a feature. 2.10.x would only get bug fixes

**Comment by potiuk** [comment]
You have to remember - this is an open-source project that you have full visibilty into the sources and if the maintainers choose not to backport something, doing that yourself is a perfectly possible thing to do .

**Comment by shalberd** [comment]
@potiuk Yes, I understand what you mean, am involved in other open source projects as well. 
Just to clarify: By backporting this feature / commit myself to v2.10.x .. do you mean I build it myself, or, that you want me to cherry pick the commit against main and apply it as a Merge Request / PR again e.g. v2-10-test branch, similar to how you handled it with another contributor in feature 42934 in the past: https://github.com/apache/airflow/pull/42934#issuecomment-2409468700

**Comment by potiuk** [comment]
> @potiuk Yes, I understand what you mean, am involved in other open source projects as well.
Just to clarify: By backporting this feature / commit myself to v2.10.x .. do you mean I build it and a custom image myself, or, that you want me to cherry pick the commit against main and apply it as a Merge Request / PR against e.g. v2-10-test branch, adding a now-missing unit test, similar to how you handled it with another contributor in feature 42934 in the past: https://github.com/apache/airflow/pull/42934#issuecomment-2409468700

I think it really depends on @ephraimbuddy and @utkarsharma2 - if he will be willing to accept such PR. I think the best way is to do it, see what it involves, get it to a green status and then ask them if that's ok to be accepted. You never know - before you do it, if it's **really easy and straightforward** - or maybe there are some hidden problems that needs to be solved.

Part of the reluctance to accept any change is the risk involved, and part of it that "accepting" the backport idea of the PR is more or less "committing to merging it" even if it will turn out that the problem is much more complex than we initially thought (happens more often than you thought). And the fact that we do not accept any more "new features" to v2 branches is not some "artifficial no new features" limitations is that we prefer to focus on 3.0 and do not want to spend any time on things that might distract us from it. This is the cost of "lost opportunity" for us.

So if you will to spend all your time to try to cherry-pick and test it and show us a green PR with potentially tests added to show that it works, and when it will turn out that it "just works" - with minimal set of changes, there is a chance - just a chance that we will accept it as backport. But you should take the risk of doing the job before knowing that it will get accepted.  It's on you - including the risk.

Other than that - you can try it yourself in your own environment. And actually trying it on your own would be  even better path because if you try on your own and it will "just work" that's even bigger proof for the RM that they can safely merge the PR - because the last thing we want to have now is more problems in v2. caused by trying to add stuff there.

**Comment by shalberd** [comment]
Full ACK, will try myself in custom build, running in own environment on top of Airflow 2.10.5 codebase.
Then, will submit a PR, agreed and OK no guarantees of it ever making it into official 2.10.x codebase. I take that risk, have taken it before in Open Data Hub / Kubeflow notebooks as well, no problem.

**Pull request merged**